fix: NetworkRigidbody authority and ownership detection

NoelStephensUnity opened the pull request:
This is a minor refactor of NetworkRigidbody and NetworkTranaform to elminate constant FixedUpdate polling in NetworkRigidBody to detect changes in ownership.
This also fixes an issue where both authoritative and non-authoritative Rigidbodies were not kinematic until spawned, now all instances are set to kinematic until authority is determined and applied (this happens during OnNetworkSpawn and OnGainedOwnership/OnLostOwnership).

[MTT-3374](https://jira.unity3d.com/browse/MTT-3374)

## Testing and Documentation
- No tests have been added.

NoelStephensUnity committed:
fix

MTT-3374
Initial changes for this fix.

NoelStephensUnity committed:
fix

Migrating the authority mode check to NetworkTransform since this is what dictates authority anyways.

NoelStephensUnity committed:
style

updated and added comments

ShadauxCat commented:
Would really like to see some tests on this so we don't have regressions of this behavior later.

NoelStephensUnity committed:
test

Adding ownership tests

NoelStephensUnity committed:
update

Realized we could further simplify NetworkRigidbody and added additional testing against owners not being Kinematic and non-owners being kinematic.
Also had to have the NetworkManager relative test prefab instances ignore each others' colliders so they wouldn't skew the position testing.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-3374-networkrigidbody-ownership-change-detection

NoelStephensUnity committed:
style

whitespace fixes

0xFA11 commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -960,5 +960,22 @@ public void Teleport(Vector3 newPosition, Quaternion newRotation, Vector3 newSca
+        /// Override this and return false to follow the owner authoritative
+        /// Otherwise, it defaults to server authoritative
+        /// </summary>
+        protected virtual bool OnIsServerAuthoritatitive()
+        {
+            return true;
+        }
+
+        /// <summary>
+        /// Determines if the <see cref="NetworkTransform"/> is server or owner authoritative.
+        /// </summary>
+        public bool IsServerAuthoritative()
+        {
+            return OnIsServerAuthoritatitive();
+        }
```
I'm actually not entirely sure about exposing these things to public (aka adding minor features) at this point, especially from the SDK but it's fine I guess.

NoelStephensUnity replied:
I could make that internal...

NoelStephensUnity committed:
style

parameter name change

NoelStephensUnity committed:
style

better clarity in a comment

NoelStephensUnity committed:
update

This fixes an issue with not being able to find ClientNetworkTransform and makes the IsServerAuhtoritative internal.

NoelStephensUnity committed:
update

realized we don't need to save the original Kinematic state.
updated a comment

NoelStephensUnity committed:
update

needed to add COM_UNITY_MODULES_PHYSICS define in runtime and exclude the test if the physics module is not included.

NoelStephensUnity committed:
style

removing 2 CR/LF

NoelStephensUnity committed:
update

didn't need all of the changes made to the asmdef file

becksebenius-unity commented on com.unity.netcode.gameobjects/Components/NetworkRigidbody.cs:
```diff
@@ -5,77 +5,96 @@ namespace Unity.Netcode.Components
     [RequireComponent(typeof(NetworkTransform))]
     public class NetworkRigidbody : NetworkBehaviour
     {
+        /// <summary>
+        /// Determines if we are server (true) or owner (false) authoritative
+        /// <see cref="NetworkTransform.IsServerAuthoritative"/>
+        /// </summary>
```
Are doc comments good practice for private members?

NoelStephensUnity replied:
Nice catch!
(removed)

becksebenius-unity commented on com.unity.netcode.gameobjects/Tests/Runtime/NetworkTransform/NetworkTransformOwnershipTests.cs:
```diff
@@ -0,0 +1,277 @@
+#if COM_UNITY_MODULES_PHYSICS
```
~~Why does this need to be here, and not elsewhere?~~ NVM - it _is_ elsewhere...

~~Also, is it needed, given that you added the physics module as a dependency to the package?~~ Also NVM, I misread the change to the package manifest

NoelStephensUnity replied:
_(sort of answering your questions in reverse here)_
It is a version define depending upon whether the physics module is included or not when test runner is running:
```
        {
            "name": "com.unity.modules.physics",
            "expression": "",
            "define": "COM_UNITY_MODULES_PHYSICS"
        }
```
So, it isn't really adding it as a dependency but just figuring out if it is available.  There are some tests (i.e. test project minimal) that do not include physics and as such things like Rigidbody won't exist.  Under this scenario and specific to what is being tested, it is actually a requirement now if you include anything from com.unity.modules.physics.
If so, then you need to wrap your integration test with #if COM_UNITY_MODULES_PHYSICS if your test depends upon it or around the areas that have dependencies depending upon the integration test(s) requirements.
_(I actually just found this out by bumping up against that wall in this PR)_

NoelStephensUnity committed:
style

removing xml doc API reference

0xFA11 commented on com.unity.netcode.gameobjects/Tests/Runtime/com.unity.netcode.runtimetests.asmdef:
```diff
@@ -13,7 +13,7 @@
         "Unity.Netcode.TestHelpers.Runtime"
     ],
-    "optionalUnityReferences": [
+     "optionalUnityReferences": [
```
no need for this extra space :P

0xFA11 commented on com.unity.netcode.gameobjects/Tests/Runtime/com.unity.netcode.runtimetests.asmdef:
```diff
@@ -31,6 +31,12 @@
         }
-    ]
-}
+    ],
+    "noEngineReferences": false
```
```suggestion
    ]
```

NoelStephensUnity committed:
Update com.unity.netcode.gameobjects/Tests/Runtime/com.unity.netcode.runtimetests.asmdef

Co-authored-by: Fatih Mar <mfatihmar@gmail.com>

NoelStephensUnity committed:
Update com.unity.netcode.runtimetests.asmdef

NoelStephensUnity committed:
fix

Had to adjust the rigid body interpolation setting to be based on authority and not NetworkTransform.Interpolate (bug from this PR)
Had to re-adjust the original rigid body test to match the NetworkRigidbody change where when it is despawned physics is not applied.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-3374-networkrigidbody-ownership-change-detection

NoelStephensUnity committed:
fix

Fixing edge case scenario with object parenting test.

TwoTenPvP reviewed (approved):
👌🏻
